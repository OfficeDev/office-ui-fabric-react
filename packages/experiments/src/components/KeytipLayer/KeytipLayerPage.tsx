import * as React from 'react';
import {
  ExampleCard,
  IComponentDemoPageProps,
  ComponentPage
} from '@uifabric/example-app-base';
import { KeytipLayerBasicExample } from './examples/KeytipLayer.Basic.Example';

const KeytipLayerBasicExampleCode = require(
  '!raw-loader!experiments/src/components/KeytipLayer/examples/KeytipLayer.Basic.Example.tsx'
) as string;

export class KeytipLayerPage extends React.Component<IComponentDemoPageProps, {}> {
  public render(): JSX.Element {
    return (
      <ComponentPage
        title='Keytip Layer'
        componentName='KeytipLayer'
        exampleCards={
          <div>
            <ExampleCard title='KeytipLayerBasic Example' code={ KeytipLayerBasicExampleCode }>
              <KeytipLayerBasicExample />
            </ExampleCard>
          </div>
        }
        overview={
          <div />
        }
        bestPractices={
          <div />
        }
        dos={
          <div>
            <ul>
              <li>@TODO Add dos</li>
            </ul>
          </div>
        }
        donts={
          <div>
            <ul>
              <li>@TODO Add dos</li>
            </ul>
          </div>
        }
        isHeaderVisible={ this.props.isHeaderVisible }
      />
    );
  }
}