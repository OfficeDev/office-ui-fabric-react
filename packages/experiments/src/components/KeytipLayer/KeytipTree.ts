import {
  IKeySequence,
  keySequencesAreEqual,
  keySequenceStartsWith,
  convertSequencesToKeytipID
} from '../../utilities/keysequence';

export interface IKeytipTreeNode {
  // ID of the <Keytip> DOM element. Needed to locate the correct keytip in the KeytipLayer's 'keytip' state array
  id: string;

  // KeySequence that invokes this KeytipTreeNode's onExecute function
  keytipSequence: IKeySequence;

  // Control's execute function for when keytip is invoked, passed from the component to the Manager in the IKeytipProps
  onExecute?: () => void;

  // Function to execute when we 'go back' to this keytip's parent.
  onReturn?: () => void;

  // List of keytips that should become visible when this keytip is pressed, can be empty
  children: string[];

  // Parent keytip
  parent: string;

  /**
   * Whether or not this node has children nodes or not. Should be used for menus/overflow components, that have
   * their children registered after the initial rendering of the DOM.
   */
  hasChildrenNodes?: boolean;

  /**
   * Optional link to another IKeytipTreeNode; only used for persisted Keytips in OverflowWells,
   * where we want keytips to be executed with and without the overflow keytip.
   */
  keytipLink?: IKeytipTreeNode;

  /**
   * Whether the keytip is visible or not in the dom.
   */
  visible?: boolean;

  // TODO: may need to know if keytip is disabled, if so shouldn't change visibility when start of sequence is pressed
}

export interface IKeytipTreeNodeMap {
  [nodeId: string]: IKeytipTreeNode;
}

export class KeytipTree {
  public currentKeytip?: IKeytipTreeNode;
  public currentSequence: IKeySequence;
  public root: IKeytipTreeNode;
  public nodeMap: IKeytipTreeNodeMap = {};

  /**
   * KeytipTree constructor
   * @param enableSequences - KeySequences that will start keytip mode, passed down through the KeytipLayer
   */
  constructor(rootId: string) {

    // Root has no keytipSequences, we instead check _enableSequences to handle multiple entry points
    this.root = {
      id: rootId,
      children: [],
      parent: '',
      keytipSequence: { keys: [] },
      hasChildrenNodes: true
    };
    this.currentSequence = { keys: [] };
    this.nodeMap[this.root.id] = this.root;
  }

  // check if it has an overflow set
  // get overflowset node
  // add node to overflowset node
  // node

  /**
   * Add a keytip node to this KeytipTree
   * @param fullSequence - Full key sequence for the keytip to add
   * @param onExecute - Callback function to trigger when this keytip is activated
   */
  public addNode(sequence: IKeySequence[], onExecute?: () => void, hasChildrenNodes?: boolean, overflowSet?: IKeySequence): void {
    let fullSequence = [...sequence];
    let nodeID = convertSequencesToKeytipID(fullSequence);
    // This keytip's sequence is the last one defined
    let keytipSequence = fullSequence.pop();
    // Parent ID is the root if there aren't any more sequences
    let parentID = fullSequence.length === 0 ? this.root.id : convertSequencesToKeytipID(fullSequence);

    // See if node already exists
    let node = this.nodeMap[nodeID];
    if (node) {
      // If node exists, it was added when one of its children was added
      // Update keytipSequence, onExecute, parent
      node.keytipSequence = keytipSequence!;
      node.onExecute = onExecute;
      node.hasChildrenNodes = hasChildrenNodes;
      node.parent = parentID;
    } else {
      // If node doesn't exist, add node
      node = {
        id: nodeID,
        keytipSequence: keytipSequence!,
        children: [],
        parent: parentID,
        onExecute,
        hasChildrenNodes,
      };
      this.nodeMap[nodeID] = node;
    }

    let parent = this.nodeMap[parentID];
    if (!parent) {
      // If parent doesn't exist, create parent with ID and children only
      parent = {
        id: parentID,
        hasChildrenNodes: true,
        children: [],
        keytipSequence: { keys: [] },
        parent: ''
      };
      this.nodeMap[parentID] = parent;
    }
    // Add node to parent's children
    parent.children.push(nodeID);
  }

  public getExactMatchedNode(keySequence: IKeySequence, currentKeytip: IKeytipTreeNode): IKeytipTreeNode | undefined {
    let possibleNodes = this._getChildrenNodes(currentKeytip.children);
    for (let node of possibleNodes) {
      if (keySequencesAreEqual(node.keytipSequence, keySequence)) {
        return node;
      }
    }
    return undefined;
  }

  public getPartiallyMatchedNodes(keySequence: IKeySequence, currentKeytip: IKeytipTreeNode): IKeytipTreeNode[] {
    let nodes: IKeytipTreeNode[] = [];
    let possibleNodes = this._getChildrenNodes(currentKeytip.children);
    for (let node of possibleNodes) {
      if (keySequenceStartsWith(node.keytipSequence, keySequence)) {
        nodes.push(node);
      }
    }
    return nodes;
  }

  private _getChildrenNodes(ids: string[]): IKeytipTreeNode[] {
    let nodes: IKeytipTreeNode[] = [];
    for (let id of ids) {
      nodes.push(this.nodeMap[id]);
    }
    return nodes;
  }

  private _getOverflowNode(overflowSequence: IKeySequence, parentSequence: IKeySequence[]): IKeytipTreeNode {
    let fullOverflowSequence = [...parentSequence, ...[overflowSequence]];
    let overflowNodeId = convertSequencesToKeytipID(fullOverflowSequence);

    let node = this.nodeMap[overflowNodeId];

    // if overflow node has not been added, we create it
    if (!node) {
      node = {
        id: overflowNodeId,
        keytipSequence: overflowSequence,
        parent: convertSequencesToKeytipID(parentSequence),
        children: [],
        hasChildrenNodes: true,
      };
      this.nodeMap[overflowNodeId] = node;
    }

    return node;
  }
}