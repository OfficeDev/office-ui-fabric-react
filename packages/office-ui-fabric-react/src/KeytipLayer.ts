export * from './packages/react-primitives/KeytipLayer';
