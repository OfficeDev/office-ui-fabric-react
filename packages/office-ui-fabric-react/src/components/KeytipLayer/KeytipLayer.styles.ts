import { ILayerStyles, ILayerStyleProps } from '../../Layer';
import { IKeytipLayerStyleProps, IKeytipLayerStyles } from './KeytipLayer.types';
import { hiddenContentStyle } from '../../Styling';

export const getLayerStyles = (props: ILayerStyleProps): ILayerStyles => {
  return {
    root: [{
      // Prioritize the Keytips above all other Layers
      zIndex: 1000001
    }]
  };
};

export const getStyles = (props: IKeytipLayerStyleProps): IKeytipLayerStyles => {
  return {
    innerContent: [{
      ...hiddenContentStyle,
      visibility: 'hidden',
      opacity: 0
    }]
  };
};