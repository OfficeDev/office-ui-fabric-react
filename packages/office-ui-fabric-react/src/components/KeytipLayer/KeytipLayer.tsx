import * as React from 'react';
import { IKeytipLayerProps } from './KeytipLayer.types';
import { getLayerStyles } from './KeytipLayer.styles';
import { Keytip, IKeytipProps } from '../../Keytip';
import { Layer } from '../../Layer';
import {
  autobind,
  BaseComponent,
  KeyCodes,
  KeytipTransitionModifier,
  IKeytipTransitionKey,
  ktpLayerId,
  ktpAriaSeparator,
  ktpAriaSeparatorId
} from '../../Utilities';
import { KeytipManager } from '../../utilities/keytips';
import { hiddenContentStyle } from '../../Styling';

export interface IKeytipLayerState {
  inKeytipMode: boolean;
  keytips: IKeytipProps[];
}

const defaultStartSequence: IKeytipTransitionKey = {
  key: 'Meta', modifierKeys: [KeytipTransitionModifier.alt]
};

const defaultExitSequence: IKeytipTransitionKey = {
  key: 'Meta', modifierKeys: [KeytipTransitionModifier.alt]
};

const defaultReturnSequence: IKeytipTransitionKey = {
  key: 'Escape'
};

/**
 * A layer that holds all keytip items
 *
 * @export
 * @class KeytipLayer
 * @extends {BaseComponent<IKeytipLayerProps>}
 */
export class KeytipLayer extends BaseComponent<IKeytipLayerProps, IKeytipLayerState> {
  public static defaultProps: IKeytipLayerProps = {
    keytipStartSequences: [defaultStartSequence],
    keytipExitSequences: [defaultExitSequence],
    keytipReturnSequences: [defaultReturnSequence],
    content: 'Alt Windows'
  };

  private _keytipManager: KeytipManager = KeytipManager.getInstance();

  // tslint:disable-next-line:no-any
  constructor(props: IKeytipLayerProps, context: any) {
    super(props, context);

    this.state = {
      inKeytipMode: false,
      // Get the initial set of keytips
      keytips: [...this._keytipManager.keytips],
    };

    this._keytipManager.init(this);
  }

  /**
   * Sets the keytips state property
   *
   * @param keytipProps - Keytips to set in this layer
   */
  public setKeytips(keytipProps: IKeytipProps[]) {
    this.setState({ keytips: keytipProps });
  }

  public render(): JSX.Element {
    const {
      content
    } = this.props;

    const {
      keytips
    } = this.state;

    const hiddenStyle = {
      ...hiddenContentStyle,
      visibility: 'hidden',
      opacity: 0
    };

    return (
      <Layer getStyles={ getLayerStyles }>
        <span id={ ktpLayerId } style={ { hiddenStyle } }>{ content }</span>
        <span id={ ktpAriaSeparatorId } style={ { hiddenStyle } }>{ ktpAriaSeparator }</span>
        { keytips && keytips.map((keytipProps: IKeytipProps, index: number) => {
          return <Keytip key={ index } { ...keytipProps } />;
        }) }
      </Layer>
    );
  }

  public componentDidMount(): void {
    // Add window listeners
    this._events.on(window, 'mousedown', this._onDismiss);
    this._events.on(window, 'resize', this._onDismiss);
    this._events.on(window, 'keydown', this._onKeyDown, true /* useCapture */);
    this._events.on(window, 'keypress', this._onKeyPress, true /* useCapture */);
    this._events.on(window, 'scroll', this._onDismiss, true /* useCapture */);
  }

  public componentWillUnmount(): void {
    // Remove window listeners
    this._events.off(window, 'mousedown', this._onDismiss);
    this._events.off(window, 'resize', this._onDismiss);
    this._events.off(window, 'keydown', this._onKeyDown, true /* useCapture */);
    this._events.off(window, 'keypress', this._onKeyPress, true /* useCapture */);
    this._events.off(window, 'scroll', this._onDismiss, true /* useCapture */);
  }

  /**
   * Exits keytip mode for this layer
   */
  public exitKeytipMode(): void {
    if (this.props.onExitKeytipMode) {
      this.props.onExitKeytipMode();
    }
    this.setState({ inKeytipMode: false });
  }

  /**
   * Enters keytip mode for this layer
   */
  public enterKeytipMode(): void {
    if (this.props.onEnterKeytipMode) {
      this.props.onEnterKeytipMode();
    }
    this.setState({ inKeytipMode: true });
  }

  @autobind
  private _onDismiss(ev?: React.MouseEvent<HTMLElement>): void {
    // if we are in keytip mode, then exit keytip mode
    if (this.state.inKeytipMode) {
      this._keytipManager.exitKeytipMode();
    }
  }

  @autobind
  private _onKeyDown(ev: React.KeyboardEvent<HTMLElement>): void {
    switch (ev.which) {
      case KeyCodes.alt:
        // ALT puts focus in the browser bar, so it should not be used as a key for keytips.
        // It can be used as a modifier
        break;
      case KeyCodes.tab:
      case KeyCodes.enter:
      case KeyCodes.space:
        this._keytipManager.exitKeytipMode();
        break;
      default:
        let key = ev.key;
        if (key === 'OS' || key === 'Win') {
          // Special cases for browser-specific changes that will be fixed in the future
          // Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1232918
          // Edge: https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/8860571/
          // and https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/16424492/
          key = 'Meta';
        }
        const transitionKey: IKeytipTransitionKey = { key };
        transitionKey.modifierKeys = this._getModifierKey(key, ev);
        this._keytipManager.processTransitionInput(transitionKey);
        break;
    }
  }

  /**
   * Gets the ModifierKeyCodes based on the keyboard event
   *
   * @param ev - React.KeyboardEvent
   * @returns List of ModifierKeyCodes that were pressed
   */
  private _getModifierKey(key: string, ev: React.KeyboardEvent<HTMLElement>): KeytipTransitionModifier[] | undefined {
    const modifierKeys = [];
    if (ev.altKey && key !== 'Alt') {
      modifierKeys.push(KeytipTransitionModifier.alt);
    }
    if (ev.ctrlKey && key !== 'Ctrl') {
      modifierKeys.push(KeytipTransitionModifier.ctrl);
    }
    if (ev.shiftKey && key !== 'Shift') {
      modifierKeys.push(KeytipTransitionModifier.shift);
    }
    if (ev.metaKey && key !== 'Meta') {
      modifierKeys.push(KeytipTransitionModifier.meta);
    }
    return modifierKeys.length ? modifierKeys : undefined;
  }

  @autobind
  private _onKeyPress(ev: React.KeyboardEvent<HTMLElement>): void {
    // Call processInput
    this._keytipManager.processInput(ev.key);
  }
}