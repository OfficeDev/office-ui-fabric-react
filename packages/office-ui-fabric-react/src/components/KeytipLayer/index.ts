export * from './KeytipLayer';
export * from './KeytipLayer.types';
export * from './KeytipManager';
export * from './KeytipTree';
export * from './KeytipUtils';