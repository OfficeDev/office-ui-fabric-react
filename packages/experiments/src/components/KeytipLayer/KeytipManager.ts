import { KeytipLayer } from './KeytipLayer';
import { KeytipTree, IKeytipTreeNode } from './KeytipTree';
import { IKeytipProps } from '../../Keytip';
import {
  IKeySequence,
  convertSequencesToKeytipID,
  keySequencesContain,
} from '../../utilities/keysequence';

export class KeytipManager {
  private static _instance: KeytipManager = new KeytipManager();

  public keytipTree: KeytipTree;
  public currentSequence: IKeySequence;

  private _layer: KeytipLayer;
  private _enableSequences: IKeySequence[];
  private _exitSequences: IKeySequence[];
  private _goBackSequences: IKeySequence[];

  /**
 * Static function to get singleton KeytipManager instance
 */
  public static getInstance(): KeytipManager {
    return this._instance;
  }

  /**
   * Sets the _layer property and creates a KeytipTree
   * Should be called from the KeytipLayer constructor
   * @param layer - KeytipLayer object
   */
  public init(layer: KeytipLayer): void {
    this._layer = layer;
    this.currentSequence = { keyCodes: [] };
    this._enableSequences = this._layer.props.keytipStartSequences!;
    this._exitSequences = this._layer.props.keytipExitSequences!;
    this._goBackSequences = this._layer.props.keytipGoBackSequences!;
    // Create the KeytipTree
    this.keytipTree = new KeytipTree(this._layer.props.id);
  }

  /**
   * Gets the aria-describedby property for a set of keySequences
   * keySequences should not include the initial keytip 'start' sequence
   * @param keySequences - Full path of IKeySequences for one keytip
   */
  public getAriaDescribedBy(keySequences: IKeySequence[]): string {
    let describedby = this._layer.props.id;
    if (!keySequences.length) {
      // Return just the layer ID
      return describedby;
    }

    for (let i = 0; i < keySequences.length; i++) {
      describedby += ' ' + convertSequencesToKeytipID(keySequences.slice(0, i + 1));
    }

    return describedby;
  }

  /**
   * Registers a keytip in _layer
   * @param keytipProps - Keytip to register
   */
  public registerKeytip(keytipProps: IKeytipProps): void {
    // Set the 'keytips' property in _layer
    this._layer && this._layer.registerKeytip(keytipProps);
  }

  public showKeytips(ids: string[]): void {
    for (let id of ids) {
      this.keytipTree.nodeMap[id] = { ...this.keytipTree.nodeMap[id], visible: true };
    }
  }

  public hideKeytips(ids?: string[]): void {
    // TODO
  }

  public getLayerId(): string {
    return this._layer.props.id;
  }

  public exitKeytipMode(): void {
    this.keytipTree.currentKeytip = undefined;
    this._layer.exitKeytipMode();
  }

  public enterKeytipMode(): void {
    this._layer.enterKeytipMode();
  }

  /**
   * Processes inputs from the document listener and traverse the keytip tree
   * @param keySequence - Keys pressed by the user
   */
  public processInput(keySequence: IKeySequence): void {
    let currentSequence = { keyCodes: [...this.currentSequence.keyCodes, ...keySequence.keyCodes] };

    // If key sequence is in 'exit sequences', exit keytip mode
    //    Trigger layer's onExit callback
    if (keySequencesContain(this._exitSequences, currentSequence) && this.keytipTree.currentKeytip) {
      this.hideKeytips();
      this.exitKeytipMode();
      return;
    }
    // If key sequence is in 'go back sequences', move currentKeytip to parent (or if currentKeytip is the root, exit)
    //    Trigger node's onGoBackExecute
    //    Hide all keytips currently showing
    //    Show all keytips of children of currentKeytip
    if (keySequencesContain(this._goBackSequences, currentSequence)) {
      if (this.keytipTree.currentKeytip) {
        if (this.keytipTree.currentKeytip.id === this.keytipTree.root.id) {
          // We are at the root, exit keytip mode
          this.exitKeytipMode();
        } else {
          // If this keytip has a goBack prop, we execute the func.
          if (this.keytipTree.currentKeytip.onGoBack) {
            this.keytipTree.currentKeytip.onGoBack();
          }

          // Return pointer to its parent
          this.keytipTree.currentKeytip = this.keytipTree.nodeMap[this.keytipTree.currentKeytip.parent!];
          this.hideKeytips(); // HIDE ALL
          this.showKeytips(this.keytipTree.currentKeytip.children);
        }
      }
      return;
    }

    // If key sequence is in 'entry sequences' and currentKeytip is null, set currentKeytip to root
    //    Show children of root
    //    Trigger layer's onEnter callback
    if (keySequencesContain(this._enableSequences, currentSequence) && !this.keytipTree.currentKeytip) {
      this.keytipTree.currentKeytip = this.keytipTree.root;
      this.showKeytips(this.keytipTree.currentKeytip.children);
      this.enterKeytipMode();
    }

    if (this.keytipTree.currentKeytip) {
      let node = this.keytipTree.getExactMatchedNode(currentSequence, this.keytipTree.currentKeytip);
      if (node) { // we found a matching node
        this.keytipTree.currentKeytip = node;
        if (this.keytipTree.currentKeytip.onExecute) {
          this.keytipTree.currentKeytip.onExecute();
        }

        if (this.keytipTree.currentKeytip.children.length === 0) {
          this.exitKeytipMode();
          // TODO: WE NEED TO CHECK IF THIS IS REALLY A LEAF OR NOT
        } else {
          this.showKeytips(this.keytipTree.currentKeytip.children);
        }
        this.hideKeytips();
        this.currentSequence = { keyCodes: [] };
        return;
      }

      let partialNodes = this.keytipTree.getPartiallyMatchedNodes(currentSequence, this.keytipTree.currentKeytip);
      if (partialNodes.length > 0) {
        // we found partial nodes, so we show only those.
        this.hideKeytips();
        let ids = partialNodes.map((partialNode: IKeytipTreeNode) => { return partialNode.id; });
        this.showKeytips(ids); // show only keytips that were partially matched
        this.currentSequence = currentSequence;
      }
    }

    // If currentKeytip is a node, look at all children of currentKeytip
    //    If the sequence exactly matches one of the children
    //      Trigger node's onExecute
    //      ** TODO: we would have to do the below after the DOM has finished rendering to know for sure if node was a leaf (e.g. menu) **
    //      If the new node is a leaf
    //        Set currentKeytip to null and exit keytip mode
    //      Else
    //        Set currentKeytip to node just triggered
    //        Hide all keytips currently showing
    //        Show all keytips of children of currentKeytip
    //    Else if the sequence matches the first part of a keytip
    //      Set visibility to false on keytips that don't match
    //    Else the sequence doesn't match anything
    //      Do nothing

    // Building up sequences
    /**
     * Have a 'currentSequence'
     * Every time processInput runs, it takes currentSequence + keySequence and tries to match that to the children of 'currentKeytip'
     * If it finds a match, it will set currentSequence += keySequence
     * When we match a whole keytip, we clear currentSequence
     */
  }
}