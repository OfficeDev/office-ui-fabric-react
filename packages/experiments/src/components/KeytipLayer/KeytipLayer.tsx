import * as React from 'react';
import { IKeytipLayerProps } from './KeytipLayer.types';
import { Keytip, IKeytipProps } from '../Keytip';
import {
  autobind,
  BaseComponent
} from '../../Utilities';
import { Layer } from 'office-ui-fabric-react/lib/Layer';
import { KeyCodes, ModifierKeyCodes } from '../../Utilities';
import { IKeytipTransitionSequence, IKeytipTransitionKey, convertSequencesToKeytipID } from '../../utilities/keysequence';
import { KeytipManager } from './KeytipManager';
import { ktpFullPrefix, ktpSeparator } from '../../utilities/keytip/KeytipUtils';

export interface IKeytipLayerState {
  inKeytipMode: boolean;
  keytips: IKeytipProps[];
}

const defaultStartSequence = {
  keys: [{ key: 'Meta', modifierKey: ModifierKeyCodes.alt }]
} as IKeytipTransitionSequence;

const defaultExitSequence = {
  keys: [{ key: 'Meta', modifierKey: ModifierKeyCodes.alt }]
} as IKeytipTransitionSequence;

const defaultGoBackSequence = {
  keys: [{ key: 'Escape' }]
} as IKeytipTransitionSequence;

/**
 * A layer that holds all keytip items
 *
 * @export
 * @class KeytipLayer
 * @extends {BaseComponent<IKeytipLayerProps>}
 */
export class KeytipLayer extends BaseComponent<IKeytipLayerProps, IKeytipLayerState> {
  public static defaultProps: IKeytipLayerProps = {
    keytipStartSequences: [defaultStartSequence],
    keytipExitSequences: [defaultExitSequence],
    keytipGoBackSequences: [defaultGoBackSequence],
    id: ktpFullPrefix + KeyCodes.alt + ktpSeparator + KeyCodes.leftWindow
  };

  private _keytipManager: KeytipManager = KeytipManager.getInstance();

  // tslint:disable-next-line:no-any
  constructor(props: IKeytipLayerProps, context: any) {
    super(props, context);

    this.state = {
      inKeytipMode: false,
      keytips: []
    };

    this._keytipManager.init(this);
  }

  public addKeytip(keytipProps: IKeytipProps): {} {
    return (previousState: IKeytipLayerState, currentProps: IKeytipLayerState) => {
      // TODO: check for duplicates
      let currentKeytips = [...previousState.keytips, ...[keytipProps]];
      return { ...previousState, keytips: currentKeytips };
    };
  }

  public setKeytipVisibility(ids: string[], visible: boolean): void {
    this.setState((previousState: IKeytipLayerState, currentProps: IKeytipLayerState) => {
      let currentKeytips: IKeytipProps[] = [...previousState.keytips];
      for (let keytip of currentKeytips) {
        let keytipId = convertSequencesToKeytipID(keytip.keySequences);
        if (ids.indexOf(keytipId) >= 0) {
          keytip.visible = visible;
        }
      }
      return { ...previousState, keytips: currentKeytips };
    });
  }

  public registerKeytip(keytipProps: IKeytipProps): void {
    this.setState(this.addKeytip(keytipProps));
  }

  public render(): JSX.Element {
    const {
      id
    } = this.props;

    const {
      keytips
    } = this.state;

    return (
      <Layer id={ id }>
        { keytips && keytips.map((keytipProps: IKeytipProps, index: number) => {
          return <Keytip key={ index } { ...keytipProps } />;
        }) }
      </Layer>
    );
  }

  public componentDidMount(): void {
    this._events.on(window, 'mousedown', this._onDismiss);
    this._events.on(window, 'resize', this._onDismiss);
    this._events.on(window, 'scroll', this._onDismiss);
    this._events.on(window, 'keydown', this._onKeyDown);
    this._events.on(window, 'keypress', this._onKeyPress);
    // TODO: SHOULD WE DO THIS: -> To remove callout when scrolled
    window.addEventListener('scroll', (): void => {
      if (this.state.inKeytipMode) {
        this._keytipManager.exitKeytipMode();
      }
    }, { capture: true });
  }

  public exitKeytipMode(): void {
    if (this.props.onExitKeytipMode) {
      this.props.onExitKeytipMode();
    }
    this.setState({ inKeytipMode: false });
  }

  public enterKeytipMode(): void {
    if (this.props.onEnterKeytipMode) {
      this.props.onEnterKeytipMode();
    }
    this.setState({ inKeytipMode: true });
  }

  @autobind
  private _onDismiss(ev?: React.MouseEvent<HTMLElement>): void {
    // if we are in keytip mode.. then exit keytip mode
    if (this.state.inKeytipMode) {
      this._keytipManager.exitKeytipMode();
    }
  }

  @autobind
  private _onKeyDown(ev: React.KeyboardEvent<HTMLElement>): void {
    switch (ev.which) {
      case KeyCodes.tab:
      case KeyCodes.enter:
      case KeyCodes.space:
        this._keytipManager.exitKeytipMode();
        break;
      default:
        // TODO: Add other meta keys
        let transitionKey: IKeytipTransitionKey = { key: ev.key };
        if (ev.altKey) {
          transitionKey.modifierKey = ModifierKeyCodes.alt;
        }
        this._keytipManager.processTransitionInput({ key: ev.key });
        break;
    }
  }

  @autobind
  private _onKeyPress(ev: React.KeyboardEvent<HTMLElement>): void {
    // call processInput
    // TODO: why don't we just pass in one key?
    this._keytipManager.processInput({ keys: [ev.key] });
  }
}