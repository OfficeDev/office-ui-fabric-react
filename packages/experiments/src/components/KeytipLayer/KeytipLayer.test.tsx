/* tslint:disable:no-unused-variable */
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { KeytipLayer } from './KeytipLayer';

const layerID = 'my-layer-id';

describe('KeytipLayer', () => {
  it('renders KeytipLayer correctly', () => {
    const componentContent = renderer.create(<KeytipLayer id={ layerID } />);
    let tree = componentContent.toJSON();
    expect(tree).toMatchSnapshot();
  });
});