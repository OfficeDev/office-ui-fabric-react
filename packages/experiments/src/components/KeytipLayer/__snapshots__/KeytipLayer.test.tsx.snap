// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeytipLayer renders KeytipLayer correctly 1`] = `
<span
  className="ms-Layer"
/>
`;
